kv/kvserver: TestReplicaLogStorage failed

kv/kvserver.TestReplicaLogStorage [failed](https://tanzanite.cluster.engflow.com/invocations/default/1ac2dcea-112b-4cae-8484-ef6b8661fbb1?testReportRun=1&testReportShard=33&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [a70010142031ac4161318994cdb806b003b48cd8](https://github.com/cockroachdb/cockroach/commits/a70010142031ac4161318994cdb806b003b48cd8):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 36884 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0047ffb20?, {0xd6a75a0, 0xc0013481e0})
	pkg/util/stop/stopper.go:227 +0x74
panic({0xae99c60?, 0xd622b50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).AssertHeld(...)
	pkg/util/syncutil/mutex_sync_race.go:130
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorage).Term(0xc0069fa420, 0x0)
	pkg/kv/kvserver/replica_raftlog.go:188 +0xb2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).termAt(0xc004c6a460, 0x0)
	pkg/kv/kvserver/replica_raftlog_test.go:231 +0xd2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).genTruncate(0xc004c6a460, 0xc00256ede0)
	pkg/kv/kvserver/replica_raftlog_test.go:283 +0x18d
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).truncateLoop(0xc004c6a460, {0xd6a75a0, 0xc0013481e0}, 0xc00256ede0)
	pkg/kv/kvserver/replica_raftlog_test.go:298 +0x3b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestReplicaLogStorage.func2({0xd6a75a0, 0xc0013481e0})
	pkg/kv/kvserver/replica_raftlog_test.go:69 +0x4d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0xbd4dfa1, 0x5})
	pkg/util/stop/stopper.go:500 +0x343
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 36839
	pkg/util/stop/stopper.go:489 +0x730
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestReplicaLogStorage
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReplicaLogStorage652795187
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_raftlog_test.go:53: seed: 7437946545422491503
*
* ERROR: a panic has occurred!
* panic: mutex is not write locked
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:791
*   | github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).AssertHeld
*   | 	pkg/util/syncutil/mutex_sync_race.go:130
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorage).Term
*   | 	pkg/kv/kvserver/replica_raftlog.go:188
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).termAt
*   | 	pkg/kv/kvserver/replica_raftlog_test.go:231
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).genTruncate
*   | 	pkg/kv/kvserver/replica_raftlog_test.go:283
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaLogStorageTest).truncateLoop
*   | 	pkg/kv/kvserver/replica_raftlog_test.go:298
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestReplicaLogStorage.func2
*   | 	pkg/kv/kvserver/replica_raftlog_test.go:69
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	pkg/util/stop/stopper.go:500
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1700
* Wraps: (2) panic: mutex is not write locked
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=33</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaLogStorage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50525

--- a/pkg/kv/kvserver/replica_raftlog_test.go
+++ b/pkg/kv/kvserver/replica_raftlog_test.go
@@ -225,7 +225,7 @@ func (rt *replicaLogStorageTest) read(rng *rand.Rand) {
 	require.Equalf(rt.t, ref, entries, "(%d,%d]", prev, last)
 }
 
-func (rt *replicaLogStorageTest) termAt(index uint64) uint64 {
+func (rt *replicaLogStorageTest) termLocked(index uint64) uint64 {
 	ref, err := rt.ms.Term(index)
 	require.NoError(rt.t, err)
 	term, err := rt.ls.Term(index)
@@ -269,21 +269,21 @@ func (rt *replicaLogStorageTest) genAppend(rng *rand.Rand) raft.StorageAppend {
 func (rt *replicaLogStorageTest) genTruncate(
 	rng *rand.Rand,
 ) (prev, next kvserverpb.RaftTruncatedState) {
+	rt.mu.Lock()
+	defer rt.mu.Unlock()
+
 	// We can only truncate entries <= committed index.
-	rt.mu.RLock()
 	committed := rt.mu.committed
 	comp := uint64(rt.ls.shMu.trunc.Index)
-	rt.mu.RUnlock()
-
 	require.Equal(rt.t, rt.ms.Compacted(), comp)
 	index := comp + rng.Uint64N(committed-comp+1)
 
 	return kvserverpb.RaftTruncatedState{
 			Index: kvpb.RaftIndex(comp),
-			Term:  kvpb.RaftTerm(rt.termAt(comp)),
+			Term:  kvpb.RaftTerm(rt.termLocked(comp)),
 		}, kvserverpb.RaftTruncatedState{
 			Index: kvpb.RaftIndex(index),
-			Term:  kvpb.RaftTerm(rt.termAt(index)),
+			Term:  kvpb.RaftTerm(rt.termLocked(index)),
 		}
 }
 

